### ISSUE TITLE
SpringCloud Gateway occasionally reports Context is empty error

### ISSUE THREAD

**Issue by michaelwangwangwang** [reactor/reactor-netty#2851]
```
java.util.NoSuchElementException: Context is empty
	at reactor.util.context.Context0.get(Context0.java:42) ~[reactor-core-3.4.16.jar!/:3.4.16]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ springfox.boot.starter.autoconfigure.SwaggerUiWebFluxConfiguration$CustomWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ org.springframework.cloud.gateway.filter.WeightCalculatorWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ org.springframework.boot.actuate.metrics.web.reactive.server.MetricsWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ HTTP POST "/callee/callee" [ExceptionHandlingWebHandler]
Original Stack Trace:
		at reactor.util.context.Context0.get(Context0.java:42) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.resources.DefaultPooledConnectionProvider$PooledConnectionAllocator.lambda$connectChannel$0(DefaultPooledConnectionProvider.java:498) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:57) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Mono.subscribe(Mono.java:4400) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Mono.subscribeWith(Mono.java:4515) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Mono.subscribe(Mono.java:4371) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Mono.subscribe(Mono.java:4307) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.internal.shaded.reactor.pool.SimpleDequePool.drainLoop(SimpleDequePool.java:437) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.netty.internal.shaded.reactor.pool.SimpleDequePool.pendingOffer(SimpleDequePool.java:600) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.netty.internal.shaded.reactor.pool.SimpleDequePool.doAcquire(SimpleDequePool.java:296) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.netty.internal.shaded.reactor.pool.AbstractPool$Borrower.request(AbstractPool.java:432) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.request(FluxContextWrite.java:136) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.resources.DefaultPooledConnectionProvider$DisposableAcquire.onSubscribe(DefaultPooledConnectionProvider.java:195) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onSubscribe(FluxContextWrite.java:101) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.internal.shaded.reactor.pool.SimpleDequePool$QueueBorrowerMono.subscribe(SimpleDequePool.java:720) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.resources.PooledConnectionProvider.lambda$acquire$2(PooledConnectionProvider.java:152) ~[reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:57) ~[reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.lambda$subscribe$0(HttpClientConnect.java:268) ~[reactor-netty-http-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:57) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.subscribe(HttpClientConnect.java:271) [reactor-netty-http-1.0.17.jar!/:1.0.17]
		at reactor.core.publisher.Flux.subscribe(Flux.java:8469) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:240) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onComplete(MonoIgnoreThen.java:203) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoNext$NextSubscriber.onComplete(MonoNext.java:102) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:83) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxIterable$IterableSubscription.fastPath(FluxIterable.java:340) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:227) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoNext$NextSubscriber.request(MonoNext.java:108) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:162) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:138) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onSubscribe(MonoIgnoreThen.java:134) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.onSubscribe(FluxPeek.java:171) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:92) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoNext$NextSubscriber.onSubscribe(MonoNext.java:70) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:165) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:87) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Mono.subscribe(Mono.java:4400) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoFromFluxOperator.subscribe(MonoFromFluxOperator.java:74) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Mono.subscribe(Mono.java:4400) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxDefaultIfEmpty$DefaultIfEmptySubscriber.onNext(FluxDefaultIfEmpty.java:101) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:295) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onNext(FluxFilterFuseable.java:337) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.MonoCollect$CollectSubscriber.onComplete(MonoCollect.java:159) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142) [reactor-core-3.4.16.jar!/:3.4.16]
		at reactor.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:400) [reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:419) [reactor-netty-core-1.0.17.jar!/:1.0.17]
		at reactor.netty.http.server.HttpServerOperations.onInboundNext(HttpServerOperations.java:600) [reactor-netty-http-1.0.17.jar!/:1.0.17]
		at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:93) [reactor-netty-core-1.0.17.jar!/:1.0.17]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at reactor.netty.http.server.HttpTrafficHandler.channelRead(HttpTrafficHandler.java:266) [reactor-netty-http-1.0.17.jar!/:1.0.17]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327) [netty-codec-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299) [netty-codec-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800) [netty-transport-classes-epoll-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:487) [netty-transport-classes-epoll-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:385) [netty-transport-classes-epoll-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.76.Final.jar!/:4.1.76.Final]
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.76.Final.jar!/:4.1.76.Final]
		at java.lang.Thread.run(Thread.java:829) [?:?]
```

**Comment by violetagg** [comment]
@michaelwangwangwang Your versions are pretty old. Please update Reactor/Spring versions and try to reproduce the issue.

**Comment by michaelwangwangwang** [comment]
I am using spingCloud 2020.0.6 version

**Comment by violetagg** [comment]
> I am using spingCloud 2020.0.6 version

This was released 1 year ago
https://github.com/spring-cloud/spring-cloud-release/wiki/Spring-Cloud-2020.0-Release-Notes

Since then we have fixes that might be relevant ...

**Comment by violetagg** [comment]
@michaelwangwangwang Looking at the file history, I believe that this exception should be fixed in version `1.0.19` with PR #2168
I'm closing this issue. Please update your Reactor Netty version and if the problem still exist we can reopen this issue.

**Comment by michaelwangwangwang** [comment]
> > I am using spingCloud 2020.0.6 version
> 
> This was released 1 year ago https://github.com/spring-cloud/spring-cloud-release/wiki/Spring-Cloud-2020.0-Release-Notes
> 
> Since then we have fixes that might be relevant ...

I don't think it has anything to do with the version, I have already upgraded to the latest version of 2020

**Comment by pderop** [comment]
Hi @michaelwangwangwang , 

you mean 2020.0.34 ? 
then can you provide a reproducer sample project using the latest version ?

thanks.

**Comment by michaelwangwangwang** [comment]
> Hi @michaelwangwangwang ,
> 
> you mean 2020.0.34 ? then can you provide a reproducer sample project using the latest version ?
> 
> thanks.

This error occurs during a microservice (not scg) rolling release

**Comment by pderop** [comment]
sorry, I got confused between projectreactor release train and spring-cloud release train.

So, what versions are you using for the following reactor netty artifacts ?

io.projectreactor.netty:reactor-netty-core
io.projectreactor.netty:reactor-netty-http

**Comment by michaelwangwangwang** [comment]
> sorry, I got confused between projectreactor release train and spring-cloud release train.
> 
> So, what versions are you using for the following reactor netty artifacts ?
> 
> io.projectreactor.netty:reactor-netty-core io.projectreactor.netty:reactor-netty-http

1.0.17

**Comment by pderop** [comment]
so, it is indeed a rather old reactor-netty version, which is part of reactor 2020.0.17 release train (March 2022).

please check last comment from Violeta: https://github.com/reactor/reactor-netty/issues/2851#issuecomment-1645034920:

> Looking at the file history, I believe that this exception should be fixed in **version 1.0.19** with PR https://github.com/reactor/reactor-netty/pull/2168

You need to update reactor-netty version and try to reproduce.
